source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+                          const uint32_t nid = this->TakeNextId();
+                          samp_inst->SetResultId(nid);
+                          postCallSI[rid] = nid;
+                          *iid = nid;
+                          bp->AddInstruction(std::move(samp_inst));
+                        }
+                      } else
+                        // reset OpSampledImage operand
+                        *iid = mapItr->second;
+                    });
+                // remember OpSampledImage in this block
+                if (spv_inst->opcode() == SpvOpSampledImage) {
+                  const uint32_t rid = spv_inst->result_id();
+                  postCallSI[rid] = rid;
+                }
+              }
+              bp->AddInstruction(std::move(spv_inst));
+            }
+            // finalize
+            newBlocks->push_back(std::move(bp));
+          } break;
+          default: {
+            // copy callee instruction and remap all input Ids
+            std::unique_ptr<ir::Instruction> spv_inst(
+                new ir::Instruction(*cpi));
+            spv_inst->ForEachInId([&callee2caller, &cpi, this](uint32_t* iid) {
+              const auto mapItr = callee2caller.find(*iid);
+              if (mapItr != callee2caller.end()) {
+                *iid = mapItr->second;
+              } else if (cpi->IsControlFlow()) {
+                const ir::Instruction* inst =
+                    def_use_mgr_->id_to_defs().find(*iid)->second;
+                if (inst->opcode() == SpvOpLabel) {
+                  // forward label reference. allocate a new label id, map it,
+                  // use
+                  // it and check for it at each label.
+                  const uint32_t nid = this->TakeNextId();
+                  callee2caller[*iid] = nid;
+                  *iid = nid;
+                }
+              }
+            });
+            // map and reset result id
+            const uint32_t rid = spv_inst->result_id();
+            if (rid != 0) {
+              const uint32_t nid = this->TakeNextId();
+              callee2caller[rid] = nid;
+              spv_inst->SetResultId(nid);
+            }
+            bp->AddInstruction(std::move(spv_inst));
+          } break;
+        }
+      });
+}
+
+bool InlinePass::Inline(ir::Function* func) {
+  bool modified = false;
+  for (auto bi = func->begin(); bi != func->end(); bi++) {
+    for (auto ii = bi->begin(); ii != bi->end();) {
+      if (ii->opcode() == SpvOp::SpvOpFunctionCall) {
+        // Inline call
+        std::vector<std::unique_ptr<ir::BasicBlock>> newBlocks;
+        std::vector<std::unique_ptr<ir::Instruction>> newVars;
+        GenInlineCode(&newBlocks, &newVars, ii, bi);
+        // update block map given replacement blocks
+        for (auto& blk : newBlocks) {
+          id2block_[blk->GetLabelId()] = &*blk;
+        }
+        // update phi functions in succesor blocks if call block
+        // is replaced with more than one block
```

Comment:
Right, the label_id of the caller blocks survives.  

Given there are other changes needed, I'd like the method comment on GenInlineCode to say that the last last block in new_blocks is the exit block of the inlined code.